lib/internal/abort_controller.js

```diff
@@ -0,0 +1,90 @@
+'use strict';
+
+// Modeled very closely on the AbortController implementation
+// in https://github.com/mysticatea/abort-controller (MIT license)
+
+const {
+  Object,
+  Symbol,
+} = primordials;
+
+const {
+  NodeEventTarget,
+  Event
+} = require('internal/event_target');
+const {
+  customInspectSymbol,
+  emitExperimentalWarning
+} = require('internal/util');
+const { inspect } = require('util');
+
+const kAborted = Symbol('kAborted');
+
+// Extends from NodeEventTarget intentionally to allow
+// the AbortSignal to be used similar to an EventEmitter.
+// This allows usage patterns like like:
+//   await once(abort.signal, 'abort')
+class AbortSignal extends NodeEventTarget {
+  get aborted() { return !!this[kAborted]; }
+
+  [customInspectSymbol](depth, options) {
+    const name = this.constructor.name;
+    if (depth < 0)
+      return name;
+
+    const opts = Object.assign({}, options, {
+      dept: options.depth === null ? null : options.depth - 1
+    });
+
+    return `${name} ${inspect({
+      aborted: this.aborted
+    }, opts)}`;
+  }
+}
+
+Object.defineProperties(AbortSignal.prototype, {
+  aborted: { enumerable: true }
+});
+
+function abortSignal(signal) {
+  if (signal[kAborted]) return;
+  signal[kAborted] = true;
+  const event = new Event('abort');
+  if (typeof signal.onabort === 'function') {
+    signal.onabort(event);
+  }
+  signal.dispatchEvent(event);
+}
+class AbortController {
+  #signal = new AbortSignal();
+
+  constructor() {
+    emitExperimentalWarning('AbortController');
+  }
+
+  get signal() { return this.#signal; }
+  abort() { abortSignal(this.#signal); }
+
+  [customInspectSymbol](depth, options) {
+    const name = this.constructor.name;
+    if (depth < 0)
+      return name;
+
+    const opts = Object.assign({}, options, {
+      dept: options.depth === null ? null : options.depth - 1
+    });
+
+    return `${name} ${inspect({
+      signal: this.signal
+    }, opts)}`;
+  }
```

Comment:
This custom inspect function is 1-to-1 identical to the other one above besides the actual property. I suggest to reuse the function and to just bind the property name to it.

I also wonder if it would make sense to add a new `optionsCustomizer` symbol to `inspect()` to define how the object should be inspected (e.g., using a different `depth`, `maxArrayLength`, `getter`, etc.). That way we could just define e.g., `[customInspectOptions] = { getter: true }` instead of having to write functions like these. We could also add another mode for getters to look like regular properties instead of indicating that it's actually a getter.
The reason for that is that we frequently use custom inspect functions that are partially complicated, require more computations and we some times miss to implement everything correct.

Reply:
Some utilities that would make this easier would be helpful, yes.